### PR TITLE
Feat/sign method

### PR DESCRIPTION
**Pull request by jeff-sexton**
### Summary
- Created new `createSignRequest` function to return a method that will add signature headers to a req object
  - Uses a modified version of `constructSignatureString` method from `fastify-hmac`
- Added `constructDigestString` method to calculate a digest header value based on a Request object body.
  - Modified from`getDigest` in `fastify-hmac`
- Added modified version of `getAlgorithm` from `fastify-hmac`
- Added tests
- Added method and usage documentation

### Example scripts (and dependencies)
- The example scripts and dependencies were removed from this PR in [this commit](https://github.com/autotelic/ietf-hmac-signatures/pull/5/commits/e875f0e07cacacebabff12a58ec372addc8ea560) as these examples and dependencies should not end up in the package release. 
- In order to test this method locally, you will either need to create your own test script and run a local instance of `fastify-hmac` or simpy temporarily revert the commit where I removed the example and follow the test plan below.
  - basic 'client' script to create and send requests to a server for development and testing
    - script uses `got` api, `cross-fetch` api, axios api and HTTP api to test `signRequest` method use
  - Basic example server from `fastify-hmac` to receive test requests from client script

### Test Plan
- clone and install `npm i`
- run dev Server using `fastify-hmac` for request verification `npm run example:basicServer`
- Create, sign and send a 'client' fetch request to the dev server by executing `example/basicRequest.js` in a separate terminal - `node example/basicReqest.js`
- On a successful request signing and server verification: 
  - The server termial will show a log with `"res":{"statusCode":200}`
  - The 'client' terminal will console log out the server response message `Response Message:  { foo: 'bar' }`
- On a failed request signing and server verification: 
  - The server termial will show a log with `"res":{"statusCode":401}`
  - The 'client' terminal will console log out the server response message `Response Message:  {
  statusCode: 401,
  error: 'Unauthorized',
  message: 'Signature verification failed'
}`

- Change request parameters in `example/basicRequest.js` and confirm requests are still verified by `fastify-hmac` example server
  - Changing and of the shared options between the server and the client will cause the verification to fail i.e. `sharedSecret`, `algorithmMap`, etc.
  - Changing the `signedHeaders`, Request body, etc. will result in a different signature that should still be verified by the example server using `fastify-hmac`.

### CODE REVIEW

File: src/createSignRequest.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+'use strict'
+const constructSignatureString = require('./constructSignatureString')
+const constructDigestString = require('./constructDigestString')
+
+const createSignRequest = (creationOptions) => {
+  const defaultOptions = {
+    constructSignatureString,
+    constructDigestString,
+    signedHeaders: ['(request-target)', '(created)'],
```
Comment (HW13) [inline src/createSignRequest.js:9]: We should probably just require the header names to passed in without the enclosing brackets.

File: src/constructSignatureString.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+const getMessage = (req, options) => {
+  const { signedHeaders, created, expires } = options
+  const { headers, method, url } = req
+  const parsedUrl = new URL(url)
+
+  return signedHeaders
+    .map((signatureHeader) => {
+      switch (signatureHeader) {
```
Comment (HW13) [inline src/constructSignatureString.js]: Could we provide compatibility with pascal case headers (ie. `Content-Type`)?
```js
 switch (signatureHeader.toLowerCase()) {
```

File: src/createSignRequest.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (signedHeaders.includes('(expires)')) {
+      combinedOptions.expires = combinedOptions.created + expiryOffset
+    }
+
+    if (signedHeaders.includes('digest')) {
+      headers.digest = `${digestAlgorithm}=${constructDigestString(req, combinedOptions)}`
+    }
+
+    headers.Signature = `keyId="${keyId}", algorithm="${algorithmName}", headers="${signedHeaders.join(' ')}", signature="${constructSignatureString(req, combinedOptions)}", created=${combinedOptions.created}${combinedOptions.expires ? `, expires=${combinedOptions.expires}` : ''}`
```
Comment (HW13) [inline src/createSignRequest.js:53]: This assumes `headers` will be a plain object, and not a `Headers` class. I have first hand experience of attempting to add headers to a `GraphQLRequest` in this manor, and having ApolloServer ignore them (needed to use `headers.set`). We could add some sort of `instanceof` logic, or change this function to return signed headers and leave it up to the user to add them to their request. Thoughts or suggestions @eadmundo @jkirkpatrick24 ?

File: src/createSignRequest.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
+'use strict'
+const constructSignatureString = require('./constructSignatureString')
+const constructDigestString = require('./constructDigestString')
+
+const createSignRequest = (creationOptions) => {
```
Comment (HW13) [inline src/createSignRequest.js:5]: Are we sure we need a creator function? None of the `creationOptions` are used outside of the return function.

File: src/createSignRequest.js (hunk 3 above)
Comment (jeff-sexton) [reply src/createSignRequest.js:53]: At first, I was passing in a Request object (which included an instance of a Headers class and a URL class) but I changed it to a simple object based on feedback from @jkirkpatrick24 . The rationale is that the input and output of the signing method is as simple as possible and we can allow the developer using the method to format the output in whatever way they need.

File: src/createSignRequest.js (hunk 4 above)
Comment (jeff-sexton) [reply src/createSignRequest.js:5]: This was a suggestion from @jkirkpatrick24 to allow the signing options to be set once and then have a reduced number of parameters for each request signed.

File: src/createSignRequest.js (hunk 1 above)
Comment (jeff-sexton) [reply src/createSignRequest.js:9]: It really depends if we see this `signedheaders` array as a representation of the `headers` field within the signature header or as a more generic object. In the `headers` field these values must have the brackets as part of the string.

For now, I added compatibility with both formats.

File: src/constructSignatureString.js (hunk 2 above)
Comment (jeff-sexton) [reply src/constructSignatureString.js]: Good idea. I added this suggestion and also formatted the header keys input by the request to all be lower case. This should cover all combinations of case mismatch.

File: src/createSignRequest.js (hunk 4 above)
Comment (HW13) [reply src/createSignRequest.js:5]: I think abstractions like that are best left up to the user. One could always wrap the function in a callback and apply their params, but we shouldn't assume everyone would prefer that.
```js
const signRequest = require('@autotelic/ietf-hmac-signatures')

module.exports = (req) => {
  return signRequest(req, {
    signedHeaders: ['request-target', 'created', 'content-length', 'custom-header'],
  })
}
```

File: src/createSignRequest.js (hunk 3 above)
Comment (HW13) [reply src/createSignRequest.js:53]: But the output is their, now mutated, request object. If we want to keep things simple, we should probably just return the signature header, and not modify the request at all.

Comment (HW13) [reply src/createSignRequest.js:53]: To clarify, I'm not saying we should just return the header or that we shouldn't mutate the request. I just think that if we are going to mutate the request, we should respect the headers' type.

Maybe we provide a `setHeader` option that defaults to:
```js
const defaultSetter = (carrier, key, value) => { carrier[key] = value }
```
Then if someone wants to use `Headers.set`:
```js
const customSetter = (carrier, key, value) => carrier.set(key, value)
```

File: src/createSignRequest.js (hunk 4 above)
Comment (jeff-sexton) [reply src/createSignRequest.js:5]: This is a good point but I think it requires more discussion before we set on a path. Is this something we can handle in an issue and as part of the discussion over @jkirkpatrick24 's spike?

We could get an initial release completed as part of this sprint and then make heavy modifications later.

File: src/createSignRequest.js (hunk 3 above)
Comment (jeff-sexton) [reply src/createSignRequest.js:53]: This is another very good point. Is this an issue and feature addition or do we need to add Header class compatibility before the initial release?

**Review by HW13 — APPROVED** [review]
Sounds like we need this in ASAP, so let's merge and talk refactor later.

**Pull request merged**